tests/collective_ops/test_allreduce.py

```diff
@@ -219,3 +219,92 @@ def foo(x):
 
     expected = 2.0
     assert jnp.array_equal(expected, res_t)
+
+
+def test_custom_vjp():
+    from mpi4jax import allreduce
+
+    # define an arbitrary functin with custom_vjp
+    @jax.custom_vjp
+    def f(x, y):
+        r = jnp.sin(x) * y
+        r = r.sum()
+        return allreduce(r, op=MPI.SUM)[0]
+
+    def f_fwd(x, y):
+        # Returns primal output and residuals to be used in backward pass by f_bwd.
+        return f(x, y), (jnp.cos(x), jnp.sin(x), y)
+
+    def f_bwd(res, g):
+        g = allreduce(g, op=MPI.SUM)[0]
+        cos_x, sin_x, y = res  # Gets residuals computed in f_fwd
+        return (cos_x * g * y, sin_x * g)
+
+    f.defvjp(f_fwd, f_bwd)
+
+    # check that it does not crash
+    _ = jax.jit(f)(jnp.ones(3), jnp.ones(3) * 2)
+    _ = jax.jit(jax.grad(f))(jnp.ones(3), jnp.ones(3) * 2)
+
+
+def test_advanced_jvp():
+    from mpi4jax import allreduce
+    from functools import partial
+
+    def expect(
+        log_pdf,
+        expected_fun,
+        pars,
+        σ,
+        *expected_fun_args,
+        n_chains,
+    ):
+        return _expect(n_chains, log_pdf, expected_fun, pars, σ, *expected_fun_args)
+
+    @partial(jax.custom_vjp, nondiff_argnums=(0, 1, 2))
+    def _expect(n_chains, log_pdf, expected_fun, pars, σ, *expected_fun_args):
+        L_σ = expected_fun(pars, σ, *expected_fun_args).reshape((n_chains, -1))
+        return allreduce(L_σ.T.mean(), op=MPI.SUM)[0] / MPI.COMM_WORLD.Get_size()
```

Comment:
`.T` does nothing here

Reply:
yes sorry, I took it out of NetKet and removed some custom structures where that stuff made sense...
I'll remove it.